modules/imgproc/src/lsd.cpp

```diff
@@ -120,10 +114,20 @@ inline bool double_equal(const double& a, const double& b)
     return (abs_diff / abs_max) <= (RELATIVE_ERROR_FACTOR * DBL_EPSILON);
 }
 
-inline bool AsmallerB_XoverY(const edge& a, const edge& b)
-{
-    if (a.p.x == b.p.x) return a.p.y < b.p.y;
-    else return a.p.x < b.p.x;
+// function to sort points by y and then by x
+inline bool AsmallerB_YoverX(const cv::Point2d &a, const cv::Point2d &b) {
+    if (a.y == b.y) return a.x < b.x;
+    else return a.y < b.y;
+}
+
+// function to get the slope of the rectangle for a specific row
+inline double get_slope(cv::Point2d p1, cv::Point2d p2) {
+    return (p2.y != p1.y) ? (p2.x - p1.x) / (p2.y - p1.y) : 0;
```

Comment:
Hi @alalek, thank you for reviewing the PR. 

The main reason for that function is that I moved the old code logic (lines below) to a function. There was a mistake because I forgot that I switched to Point2d, so I changed the check to use integer values instead.
https://github.com/opencv/opencv/blob/cb2052dbfef821f8fbdcc7ecf5780ab712d4e5dc/modules/imgproc/src/lsd.cpp#L1024-L1032

But the reality is that due to the way the vertices are ordered (the image below shows the possible cases), and the way the 2 regions are defined for the slope calculation, the y value of both points should only be the same when the rectangle is horizontal. In that case, in the calculation for the second left region (vertices 1 and 2), and the first right region (vertices 0 and 3) the y value of the points used for the slope calculation will be the same.

And because of the checks inside the for loop (code below), the slope in those cases will not be used. So, in theory, it does not matter what the slope is when the y value is the same. But actually, returning 0 is a safe option, in case there is a mistake (like the one I just fixed with the left limit check).
https://github.com/opencv/opencv/blob/cec7cc4162925ea4afe7da6c9a8684190eb15d74/modules/imgproc/src/lsd.cpp#L988-L996

However, despite all that (sorry for the long response btw), I can make other changes to the function if necessary.